src/main/java/org/truffleruby/core/hash/HashNodes.java

```diff
@@ -900,14 +896,37 @@ protected RubyHash rehashBuckets(RubyHash hash,
 
                 if (bucketEntry == null) {
                     entries[index] = entry;
+                    previousEntry = entry;
                 } else {
+                    boolean encounteredDuplicateKey = false;
                     while (bucketEntry.getNextInLookup() != null) {
+                        if (compareHashKeysNode.equalKeys(
+                                compareByIdentity,
+                                entry.getKey(),
+                                entry.getHashed(),
+                                bucketEntry.getKey(),
+                                bucketEntry.getHashed())) {
+                            encounteredDuplicateKey = true;
+                            break;
+                        }
                         bucketEntry = bucketEntry.getNextInLookup();
                     }
+                    if (encounteredDuplicateKey || compareHashKeysNode.equalKeys(
+                            compareByIdentity,
+                            entry.getKey(),
+                            entry.getHashed(),
+                            bucketEntry.getKey(),
+                            bucketEntry.getHashed())) { // If the bucket contains a single entry, we never set the flag
+                        if (previousEntry != null) {
+                            previousEntry.setNextInSequence(entry.getNextInSequence());
```

Comment:
What about `entry.getNextInSequence().getPreviousInSequence()`, won't that still refer to `entry`?

Reply:
I think we should use `BucketsStrategy.removeFromSequenceChain` here, then it should be fine.